core/common/src/main/java/alluxio/underfs/ObjectLowLevelOutputStream.java

```diff
@@ -0,0 +1,402 @@
+  }
+
+  protected void uploadPart(File file, int partNumber, boolean lastPart) throws IOException {
+    final String md5;
+    if (mHash != null) {
+      md5 = Base64.encodeBase64String(mHash.digest());
+    } else {
+      md5 = null;
+    }
+    Callable<?> callable = () -> {
+      try {
+        RetryUtils.retry("upload part for key " + mKey + " and part number " + partNumber,
+            () -> uploadPartInternal(file, partNumber, lastPart, md5), mRetryPolicy);
+        return null;
+      } finally {
+        // Delete the uploaded or failed to upload file
+        if (!file.delete()) {
+          LOG.error("Failed to delete temporary file @ {}", file.getPath());
+        }
+      }
+    };
+    ListenableFuture<?> futureTag = mExecutor.submit(callable);
+    mFutures.add(futureTag);
+    LOG.debug(
+        "Submit upload part request. key={}, partNum={}, file={}, fileSize={}, lastPart={}.",
+        mKey, partNumber, file.getPath(), file.length(), lastPart);
+  }
+
+  protected void abortMultiPartUpload() throws IOException {
+    RetryUtils.retry("abort multipart upload for key " + mKey, this::abortMultiPartUploadInternal,
+        mRetryPolicy);
+  }
+
+  protected void waitForAllPartsUpload() throws IOException {
+    try {
+      for (ListenableFuture<?> future : mFutures) {
+        if (mUploadPartTimeout == null) {
+          future.get();
+        } else {
+          future.get(mUploadPartTimeout.toMillis(), TimeUnit.MILLISECONDS);
+        }
+      }
+    } catch (ExecutionException e) {
+      // No recover ways so that we need to cancel all the upload tasks
+      // and abort the multipart upload
+      Futures.allAsList(mFutures).cancel(true);
+      abortMultiPartUpload();
+      throw new IOException(
+          "Part upload failed in multipart upload with to " + mKey, e);
+    } catch (InterruptedException e) {
+      LOG.warn("Interrupted object upload.", e);
+      Futures.allAsList(mFutures).cancel(true);
+      abortMultiPartUpload();
+      Thread.currentThread().interrupt();
+    } catch (TimeoutException e) {
+      LOG.error("timeout when upload part");
+      Futures.allAsList(mFutures).cancel(true);
+      abortMultiPartUpload();
+      throw new IOException("timeout when upload part " + mKey, e);
+    }
+    mFutures.clear();
+  }
+
+  protected abstract void uploadPartInternal(
+      File file,
+      int partNumber,
+      boolean isLastPart,
+      @Nullable String md5)
+      throws IOException;
+
+  protected abstract void initMultiPartUploadInternal() throws IOException;
+
+  protected abstract void completeMultiPartUploadInternal() throws IOException;
+
+  protected abstract void abortMultiPartUploadInternal() throws IOException;
+
+  protected abstract void createEmptyObject(String key) throws IOException;
+
+  protected abstract void putObject(String key, File file, String md5) throws IOException;
+
+  protected abstract boolean isMultiPartUploadInitialized();
```

Comment:
can put mUploadId in this file and directly check the value? without each subfile to overwrite the method?